gui.py

```diff
@@ -235,10 +294,55 @@ def initUI(self, song):
 
         self.update()
 
+    def updatePorts(self):
+
+        self.ports_initialised.clear()
+        client = self._jack_client
+        print("Outputs ports : ")
+        for port in client.outports:
+            print(port)
+
+        self.port_names_by_output = {o: list(self.song.channel_names(o)) for o
+                                     in self.song.outputs}
+
+        wanted_ports = set(
+            c for os in self.port_names_by_output.values() for c in os)
+
+        print("wanted ports:")
+        print(wanted_ports)
+
+        # remove unwanted ports
+        for port in [port for port in client.outports
+                     if port.shortname not in wanted_ports | set(
+                Clip.default_outports())]:
+            port.unregister()
+
+        print("Outputs after removing unused ports: ")
+        for port in client.outports:
+            print(port)
+
+        # create new ports
+        current_ports = set(port.shortname for port in client.outports)
+        for new_port_name in wanted_ports - current_ports:
+            client.outports.register(new_port_name)
+
+        print("Outputs after adding new ports: ")
+        for port in client.outports:
+            print(port)
+
+        self.output_ports_by_name = {port.shortname: port for port in
+                                     client.outports}
+        print(self.output_ports_by_name)
+        self.ports_initialised.set()
+
     def closeEvent(self, event):
         settings = QSettings('superboucle', 'devices')
         settings.setValue('devices',
                           [pickle.dumps(x.mapping) for x in self.devices])
+        self.settings.setValue('playlist',
```

Comment:
Ok , I did not notice `self.` prefix ! About QSettings, it handle multiple platform, this is why I choose it. For data, I use pickle because it handle tuple storage, json convert it to list.